internal/service/cluster/discovery/join_peers.go

```diff
@@ -0,0 +1,178 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/samber/lo"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/grafana/alloy/internal/runtime/logging/level"
+)
+
+// newWithJoinPeers creates a DiscoverFn that resolves the provided list of peers to a list of addresses that can be
+// used for clustering. The peers can be given in the following formats:
+//   - IP address, e.g. 10.10.10.10, in which case the default port is used
+//   - IP address with port, e.g. 10.10.10.10:12345, in which case the provided port is used
+//   - Hostname, e.g. example.com - in this case the SRV records are looked up for the hostname and all the resolved
+//     IP addresses are used with the default port.
+//   - Hostname with port, e.g. example.com:12345 - in this case the SRV records are looked up for the hostname and all
+//     the resolved IP addresses will be used with the provided port.
+//
+// TODO(thampiotr): Update this after adding A record support and probably move this to documentation.
+func newWithJoinPeers(opts Options) DiscoverFn {
+	return func() ([]string, error) {
+		ctx, span := opts.Tracer.Tracer("").Start(
+			context.Background(),
+			"ResolveClusterJoinAddresses",
+			trace.WithSpanKind(trace.SpanKindInternal),
+			trace.WithAttributes(attribute.Int("join_peers_count", len(opts.JoinPeers))),
+		)
+		defer span.End()
+
+		// Use these resolvers in order to resolve the provided addresses into a form that can be used by clustering.
+		resolvers := []addressResolver{
+			ipResolver(opts.Logger),
+			dnsAResolver(opts, ctx),
+			dnsSRVResolver(opts, ctx),
+		}
+
+		// Get the addresses.
+		addresses, err := buildJoinAddresses(opts, resolvers)
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+			return nil, fmt.Errorf("static peer discovery: %w", err)
+		}
+
+		// Return unique addresses.
+		result := lo.Uniq(addresses)
+		span.SetAttributes(attribute.Int("resolved_addresses_count", len(result)))
+		span.SetStatus(codes.Ok, "resolved addresses")
+		return result, nil
+	}
+}
+
+func buildJoinAddresses(opts Options, resolvers []addressResolver) ([]string, error) {
+	var (
+		result      []string
+		deferredErr error
+	)
+
+	for _, addr := range opts.JoinPeers {
+		// See if we have a port override, if not use the default port.
+		host, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			host = addr
+			port = strconv.Itoa(opts.DefaultPort)
+		}
+
+		atLeastOneSuccess := false
+		for _, resolver := range resolvers {
+			resolved, err := resolver(host)
+			deferredErr = errors.Join(deferredErr, err)
+			for _, foundAddr := range resolved {
+				result = append(result, appendPortIfAbsent(foundAddr, port))
```

Comment:
Do we need this? Is this always handled in the above code starting on line 69.

Reply:
You're right, can be simplified to `net.JoinHostPort(foundAddr, port)`.